### PR TITLE
Added description to list of metrics reported by cni-metrics-helper

### PR DESCRIPTION
**Pull request by zachdorame**
**What type of PR is this?**

documentation

**Which issue does this PR fix**:
https://github.com/aws/amazon-vpc-cni-k8s/issues/2421

**What does this PR do / Why do we need it**:
This PR provides a description of each metric reported by `cni-metrics-helper`. It consolidates both the metric name and metric description into a table.

**If an issue # is not available please add repro steps and logs from IPAMD/CNI showing the issue**:

**Testing done on this change**:

N/A

**Will this PR introduce any new dependencies?**:

No

**Will this break upgrades or downgrades? Has updating a running cluster been tested?**:
No, No

**Does this change require updates to the CNI daemonset config files to work?**:

No

**Does this PR introduce any user-facing change?**:

No

```release-note

```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by zachdorame** [comment]
In https://github.com/aws/amazon-vpc-cni-k8s/issues/2421#issue-1756923522 there was an ask of providing a description of _"how to use [the metrics] to monitor the health of the cluster."_ I think the description of each metric also satisfies this, but please let me know if it doesn't

**Review by jdn5126 — COMMENTED** [review]
This is a great start, thanks @zachdorame! I believe the full list of metrics we support can be found at https://github.com/aws/amazon-vpc-cni-k8s/blob/master/cmd/cni-metrics-helper/metrics/cni_metrics.go#L30

File: cmd/cni-metrics-helper/README.md [hunk 1]
```diff
@@ -15,24 +15,25 @@ The following diagram shows how `cni-metrics-helper` works in a cluster:
...
-"podENIErr",
-"reconcileCount",
-"totalIPAddresses",
-"totalIPv4Prefixes",
-"totalAssignedIPv4sPerCidr"
-```
+
+| Metric | Description |
+| ------ | ----------- |
+| addReqCount | The number of IP address requests |
```
Comment (jdn5126) [inline cmd/cni-metrics-helper/README.md:21]: For each metric, we should mention that it is an aggregate of the count across all `aws-node` pods. So for this one, we can reference https://github.com/aws/amazon-vpc-cni-k8s/blob/master/cmd/cni-metrics-helper/metrics/cni_metrics.go#L129, which points us to https://github.com/aws/amazon-vpc-cni-k8s/blob/master/utils/prometheusmetrics/prometheusmetrics.go#L64

And then the use is here: https://github.com/aws/amazon-vpc-cni-k8s/blob/master/pkg/ipamd/rpc_handler.go#L70

So we can have a description like: "Aggregate of the number of CNI ADD requests that require an IP address."

Comment (zachdorame) [reply cmd/cni-metrics-helper/README.md]: Got it, I found that all of the metrics except for `awsAPILatency` were aggregate sums which was just the maximum data point. I changed the wording in the table to reflect that

File: cmd/cni-metrics-helper/README.md [hunk 2]
```diff
@@ -15,24 +15,29 @@ The following diagram shows how `cni-metrics-helper` works in a cluster:
...
+| Metric | Description |
+| ------ | ----------- |
+| addReqCount | Aggregate of the number of CNI ADD requests that require an IP address |
+| assignIPAddresses | Aggregate of the number of IP addresses assigned to pods |
+| awsAPIErr | Aggregate of the number of times AWS API returns an error |
+| awsAPILatency | The maximum of AWS API call latency in ms |
+| awsUtilErr | Aggregate of the number of errors not handled in awsutils library |
+| delReqCount | Aggregate of the number of delete IP address requests |
+| eniAllocated | Aggregate of the number of ENIs allocated |
+| eniMaxAvailable | Aggregate of the maximum number of ENIs that can be attached to this instance, accounting for unmanaged ENIs |
```
Comment (jdn5126) [inline cmd/cni-metrics-helper/README.md:28]: This is a strange metric. Maybe reword this as: "This metric sums the maximum number of ENIs that can be attached to *each* instance, accounting for unmanaged ENIs."

File: cmd/cni-metrics-helper/README.md [hunk 3]
```diff
@@ -15,24 +15,29 @@ The following diagram shows how `cni-metrics-helper` works in a cluster:
...
+| ipamdActionInProgress | Aggregate of the number of ipamd actions in progress |
+| ipamdErr | Aggregate of the number of errors encountered in ipamd |
+| maxIPAddresses | Aggregate of the maximum number of IP addresses that can be allocated to the instance |
```
Comment (jdn5126) [inline cmd/cni-metrics-helper/README.md:31]: Same as the ENI one, this can be: "This metric sums the maximum number of IPs that can be allocated to each instance."

File: cmd/cni-metrics-helper/README.md [hunk 4]
```diff
@@ -15,24 +15,29 @@ The following diagram shows how `cni-metrics-helper` works in a cluster:
...
+| podENIErr | Aggregate of the number of errors encountere for pod ENIs |
```
Comment (jdn5126) [inline cmd/cni-metrics-helper/README.md:32]: This can be reworded as: "The number of errors encountered while managing ENIs for pods, summed across all nodes."

File: cmd/cni-metrics-helper/README.md [hunk 5]
```diff
@@ -15,24 +15,29 @@ The following diagram shows how `cni-metrics-helper` works in a cluster:
...
+| reconcileCount | Aggregate of the number of times ipamd reconciles on ENIs and IP/Prefix addresses |
+| totalIPAddresses | Aggregate of the total number of IP addresses |
```
Comment (jdn5126) [inline cmd/cni-metrics-helper/README.md:34]: nit: "This metric sums the total number of IPs that are allocated for pods across all nodes."

**Review by jdn5126 — COMMENTED** [review]
This is getting there, now we can just focus on the wording of the individual metrics. Let's try to use consistent phrasing for each. I know the phrasing in my suggestions is a bit off.

File: cmd/cni-metrics-helper/README.md [hunk 6]
```diff
@@ -15,24 +15,29 @@ The following diagram shows how `cni-metrics-helper` works in a cluster:
...
-"reconcileCount",
-"totalIPAddresses",
-"totalIPv4Prefixes",
-"totalAssignedIPv4sPerCidr"
-```
+
+| Metric | Description |
+| ------ | ----------- |
+| addReqCount | Aggregate of the number of CNI ADD requests that require an IP address |
+| assignIPAddresses | Aggregate of the number of IP addresses assigned to pods |
```
[the remark this answers is not in the document]
Comment (zachdorame) [reply cmd/cni-metrics-helper/README.md:22]: I'll adjust the wording, and I'm also thinking it might help reading to do away with specifying that a metric is an aggregate, and instead just add another column so the chart may be read like...

| Metric | Description | Representation |
| ------ | ------------ | --------------- |
| fooMetric | ... | Aggregate across all nodes |
| barMetric | ... | Maximum datapoint across all nodes | 

Something like this, what do you think?

[the remark this answers is not in the document]
Comment (jdn5126) [reply cmd/cni-metrics-helper/README.md:22]: I like that, that's a great call! Instead of "Representation", maybe we can just have "Summation/Maximum", since that is all we support?

**Review by jdn5126 — APPROVED** [review]
Looks great, thanks for doing this!
